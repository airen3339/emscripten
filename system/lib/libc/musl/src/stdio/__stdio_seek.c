#include "stdio_impl.h"

off_t __stdio_seek(FILE *f, off_t off, int whence)
{
	off_t ret;
#ifdef SYS__llseek
	off_t off_hi =
#ifndef __EMSCRIPTEN__
		off >> 32;
#else
		0;
#endif
	if (syscall(SYS__llseek, f->fd, off_hi, off, &ret, whence)<0)
		ret = -1;
#else
	ret = syscall(SYS_lseek, f->fd, off, whence);
#endif
	return ret;
}
