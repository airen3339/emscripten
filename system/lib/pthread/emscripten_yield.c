#include <stdbool.h>
#include <sched.h>
#include <threads.h>
#include "syscall.h"

#include <emscripten/threading.h>

static _Atomic bool thread_crashed = false;

void _emscripten_thread_crashed() {
  thread_crashed = true;
}

static void dummy()
{
}

static void dummy2(double now)
{
}

weak_alias(dummy, _emscripten_thread_sync_code);
weak_alias(dummy2, _emscripten_check_timers);

void _emscripten_yield(double now) {
  int is_runtime_thread = emscripten_is_main_runtime_thread();

  // When a secondary thread crashes, we need to be able to interrupt the main
  // thread even if it's in a blocking/looping on a mutex.  We want to avoid
  // using the normal proxying mechanism to send this message since it can
  // allocate (or otherwise itself crash) so use a low level atomic primitive
  // for this signal.
  if (is_runtime_thread) {
    if (thread_crashed) {
      // Return the event loop so we can handle the message from the crashed
      // thread.
      emscripten_unwind_to_js_event_loop();
    }

    // This is no-op in programs that don't include use of itimer/alarm.
    _emscripten_check_timers(now);

    // Assist other threads by executing proxied operations that are effectively
    // singlethreaded.
    emscripten_main_thread_process_queued_calls();
  }

  _emscripten_thread_sync_code();
}
