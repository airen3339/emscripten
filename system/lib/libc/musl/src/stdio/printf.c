#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int iprintf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vfiprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vfprintf(FILE *restrict f, const char *restrict fmt, va_list ap);

int __small_printf(const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vfprintf(stdout, fmt, ap);
	va_end(ap);
	return ret;
}

