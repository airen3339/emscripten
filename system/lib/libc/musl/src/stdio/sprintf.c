#include "stdio_impl.h"
#include <stdio.h>
#include <stdarg.h>

int sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

// XXX EMSCRIPTEN
int siprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = vsiprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

extern int __small_vsprintf(char *restrict s, const char *restrict fmt, va_list ap);

int __small_sprintf(char *restrict s, const char *restrict fmt, ...)
{
	int ret;
	va_list ap;
	va_start(ap, fmt);
	ret = __small_vsprintf(s, fmt, ap);
	va_end(ap);
	return ret;
}

