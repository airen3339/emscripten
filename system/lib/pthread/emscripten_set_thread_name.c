/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */
#include <string.h>

#include <emscripten/proxying.h>

#include "threading_internal.h"
#include "pthread_impl.h"

#ifndef NDEBUG
// JS function for setting thread name within devtools.
void _emscripten_set_js_thread_name(const char* name);

static void set_js_thread_name(void* arg) {
  const char* name = (const char*)arg;
  _emscripten_set_js_thread_name(name);
}
#endif

void emscripten_set_thread_name(pthread_t thread, const char* name) {
  if (thread->profilerBlock) {
    strncpy(thread->profilerBlock->name, name, EM_THREAD_NAME_MAX-1);
  }
#ifndef NDEBUG
  if (thread == pthread_self()) {
    _emscripten_set_js_thread_name(name);
  } else {
    emscripten_proxy_sync(emscripten_proxy_get_system_queue(),
                          thread,
                          set_js_thread_name,
                          (void*)name);
  }
#endif
}
