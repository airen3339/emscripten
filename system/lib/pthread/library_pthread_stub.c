#include <errno.h>
#include <pthread.h>
#include <stdint.h>
#include <stdlib.h>

int emscripten_has_threading_support() { return 0; }

int emscripten_num_logical_cores() { return 1; }

void emscripten_force_num_logical_cores(int cores) {
  // no-op, in singlethreaded builds we will always report exactly one core.
}

uint8_t emscripten_atomic_exchange_u8(void /*uint8_t*/* addr, uint8_t newVal) {
  uint8_t old = *(uint8_t*)addr;
  *(uint8_t*)addr = newVal;
  return old;
}

uint16_t emscripten_atomic_exchange_u16(void /*uint16_t*/* addr, uint16_t newVal) {
  uint16_t old = *(uint16_t*)addr;
  *(uint16_t*)addr = newVal;
  return old;
}

uint32_t emscripten_atomic_exchange_u32(void /*uint32_t*/* addr, uint32_t newVal) {
  uint32_t old = *(uint32_t*)addr;
  *(uint32_t*)addr = newVal;
  return old;
}

uint64_t emscripten_atomic_exchange_u64(void /*uint64_t*/* addr, uint64_t newVal) {
  uint64_t old = *(uint64_t*)addr;
  *(uint64_t*)addr = newVal;
  return old;
}

uint8_t emscripten_atomic_cas_u8(void /*uint8_t*/* addr, uint8_t oldVal, uint8_t newVal) {
  uint8_t old = *(uint8_t*)addr;
  if (old == oldVal)
    *(uint8_t*)addr = newVal;
  return old;
}

uint16_t emscripten_atomic_cas_u16(void /*uint16_t*/* addr, uint16_t oldVal, uint16_t newVal) {
  uint16_t old = *(uint16_t*)addr;
  if (old == oldVal)
    *(uint16_t*)addr = newVal;
  return old;
}

uint32_t emscripten_atomic_cas_u32(void /*uint32_t*/* addr, uint32_t oldVal, uint32_t newVal) {
  uint32_t old = *(uint32_t*)addr;
  if (old == oldVal)
    *(uint32_t*)addr = newVal;
  return old;
}

uint64_t emscripten_atomic_cas_u64(void /*uint64_t*/* addr, uint64_t oldVal, uint64_t newVal) {
  uint64_t old = *(uint64_t*)addr;
  if (old == oldVal)
    *(uint64_t*)addr = newVal;
  return old;
}

uint8_t emscripten_atomic_load_u8(const void /*uint8_t*/* addr) { return *(uint8_t*)addr; }

uint16_t emscripten_atomic_load_u16(const void /*uint16_t*/* addr) { return *(uint16_t*)addr; }

uint32_t emscripten_atomic_load_u32(const void /*uint32_t*/* addr) { return *(uint32_t*)addr; }

float emscripten_atomic_load_f32(const void /*float*/* addr) { return *(float*)addr; }

uint64_t emscripten_atomic_load_u64(const void /*uint64_t*/* addr) { return *(uint64_t*)addr; }

double emscripten_atomic_load_f64(const void /*double*/* addr) { return *(double*)addr; }

uint8_t emscripten_atomic_store_u8(void /*uint8_t*/* addr, uint8_t val) {
  return *(uint8_t*)addr = val;
}

uint16_t emscripten_atomic_store_u16(void /*uint16_t*/* addr, uint16_t val) {
  return *(uint16_t*)addr = val;
}

uint32_t emscripten_atomic_store_u32(void /*uint32_t*/* addr, uint32_t val) {
  return *(uint32_t*)addr = val;
}

float emscripten_atomic_store_f32(void /*float*/* addr, float val) { return *(float*)addr = val; }

uint64_t emscripten_atomic_store_u64(void /*uint64_t*/* addr, uint64_t val) {
  return *(uint64_t*)addr = val;
}

double emscripten_atomic_store_f64(void /*double*/* addr, double val) {
  return *(double*)addr = val;
}

void emscripten_atomic_fence() {
  // nop
}

uint8_t emscripten_atomic_add_u8(void /*uint8_t*/* addr, uint8_t val) {
  uint8_t old = *(uint8_t*)addr;
  *(uint8_t*)addr = old + val;
  return old;
}

uint16_t emscripten_atomic_add_u16(void /*uint16_t*/* addr, uint16_t val) {
  uint16_t old = *(uint16_t*)addr;
  *(uint16_t*)addr = old + val;
  return old;
}

uint32_t emscripten_atomic_add_u32(void /*uint32_t*/* addr, uint32_t val) {
  uint32_t old = *(uint32_t*)addr;
  *(uint32_t*)addr = old + val;
  return old;
}

uint64_t emscripten_atomic_add_u64(void /*uint64_t*/* addr, uint64_t val) {
  uint64_t old = *(uint64_t*)addr;
  *(uint64_t*)addr = old + val;
  return old;
}

uint8_t emscripten_atomic_sub_u8(void /*uint8_t*/* addr, uint8_t val) {
  uint8_t old = *(uint8_t*)addr;
  *(uint8_t*)addr = old - val;
  return old;
}

uint16_t emscripten_atomic_sub_u16(void /*uint16_t*/* addr, uint16_t val) {
  uint16_t old = *(uint16_t*)addr;
  *(uint16_t*)addr = old - val;
  return old;
}

uint32_t emscripten_atomic_sub_u32(void /*uint32_t*/* addr, uint32_t val) {
  uint32_t old = *(uint32_t*)addr;
  *(uint32_t*)addr = old - val;
  return old;
}

uint64_t emscripten_atomic_sub_u64(void /*uint64_t*/* addr, uint64_t val) {
  uint64_t old = *(uint64_t*)addr;
  *(uint64_t*)addr = old - val;
  return old;
}

uint8_t emscripten_atomic_and_u8(void /*uint8_t*/* addr, uint8_t val) {
  uint8_t old = *(uint8_t*)addr;
  *(uint8_t*)addr = old & val;
  return old;
}

uint16_t emscripten_atomic_and_u16(void /*uint16_t*/* addr, uint16_t val) {
  uint16_t old = *(uint16_t*)addr;
  *(uint16_t*)addr = old & val;
  return old;
}

uint32_t emscripten_atomic_and_u32(void /*uint32_t*/* addr, uint32_t val) {
  uint32_t old = *(uint32_t*)addr;
  *(uint32_t*)addr = old & val;
  return old;
}

uint64_t emscripten_atomic_and_u64(void /*uint64_t*/* addr, uint64_t val) {
  uint64_t old = *(uint64_t*)addr;
  *(uint64_t*)addr = old & val;
  return old;
}

uint8_t emscripten_atomic_or_u8(void /*uint8_t*/* addr, uint8_t val) {
  uint8_t old = *(uint8_t*)addr;
  *(uint8_t*)addr = old | val;
  return old;
}

uint16_t emscripten_atomic_or_u16(void /*uint16_t*/* addr, uint16_t val) {
  uint16_t old = *(uint16_t*)addr;
  *(uint16_t*)addr = old | val;
  return old;
}

uint32_t emscripten_atomic_or_u32(void /*uint32_t*/* addr, uint32_t val) {
  uint32_t old = *(uint32_t*)addr;
  *(uint32_t*)addr = old | val;
  return old;
}

uint64_t emscripten_atomic_or_u64(void /*uint64_t*/* addr, uint64_t val) {
  uint64_t old = *(uint64_t*)addr;
  *(uint64_t*)addr = old | val;
  return old;
}

uint8_t emscripten_atomic_xor_u8(void /*uint8_t*/* addr, uint8_t val) {
  uint8_t old = *(uint8_t*)addr;
  *(uint8_t*)addr = old ^ val;
  return old;
}

uint16_t emscripten_atomic_xor_u16(void /*uint16_t*/* addr, uint16_t val) {
  uint16_t old = *(uint16_t*)addr;
  *(uint16_t*)addr = old ^ val;
  return old;
}

uint32_t emscripten_atomic_xor_u32(void /*uint32_t*/* addr, uint32_t val) {
  uint32_t old = *(uint32_t*)addr;
  *(uint32_t*)addr = old ^ val;
  return old;
}

uint64_t emscripten_atomic_xor_u64(void /*uint64_t*/* addr, uint64_t val) {
  uint64_t old = *(uint64_t*)addr;
  *(uint64_t*)addr = old ^ val;
  return old;
}

uint64_t _emscripten_atomic_fetch_and_add_u64(void* addr, uint64_t val) {
  return emscripten_atomic_add_u64(addr, val);
}

uint64_t _emscripten_atomic_fetch_and_sub_u64(void* addr, uint64_t val) {
  return emscripten_atomic_sub_u64(addr, val);
}

uint64_t _emscripten_atomic_fetch_and_and_u64(void* addr, uint64_t val) {
  return emscripten_atomic_and_u64(addr, val);
}

uint64_t _emscripten_atomic_fetch_and_or_u64(void* addr, uint64_t val) {
  return emscripten_atomic_or_u64(addr, val);
}

uint64_t _emscripten_atomic_fetch_and_xor_u64(void* addr, uint64_t val) {
  return emscripten_atomic_xor_u64(addr, val);
}

int emscripten_futex_wait(
  volatile void /*uint32_t*/* addr, uint32_t val, double maxWaitMilliseconds) {
  // nop
  return 0; // success
}

int emscripten_futex_wake(volatile void /*uint32_t*/* addr, int count) {
  // nop
  return 0; // success
}

int emscripten_is_main_runtime_thread() { return 1; }

void emscripten_main_thread_process_queued_calls() {
  // nop
}

void emscripten_current_thread_process_queued_calls() {
  // nop
}

int pthread_mutex_init(
  pthread_mutex_t* __restrict mutex, const pthread_mutexattr_t* __restrict attr) {
  return 0;
}

#ifndef EMSCRIPTEN_ASAN_OVERRIDE
int pthread_mutex_lock(pthread_mutex_t* mutex) { return 0; }

int pthread_mutex_unlock(pthread_mutex_t* mutex) { return 0; }
#endif

int pthread_mutex_trylock(pthread_mutex_t* mutex) { return 0; }

struct timespec;

int pthread_mutex_timedlock(
  pthread_mutex_t* __restrict mutex, const struct timespec* __restrict t) {
  return 0;
}

int pthread_mutex_destroy(pthread_mutex_t* mutex) { return 0; }

int pthread_mutex_consistent(pthread_mutex_t* mutex) { return 0; }

int pthread_barrier_init(
  pthread_barrier_t* __restrict mutex, const pthread_barrierattr_t* __restrict attr, unsigned u) {
  return 0;
}

int pthread_barrier_destroy(pthread_barrier_t* mutex) { return 0; }

int pthread_barrier_wait(pthread_barrier_t* mutex) { return 0; }

// pthread_key_t is 32-bit, so to be able to store pointers in there, we sadly
// have to track an array of them.
static size_t num_tls_entries = 0;
static size_t max_tls_entries = 0;
struct entry_t { const void* value; int allocated; };
static struct entry_t* tls_entries = NULL;

int pthread_key_create(pthread_key_t* key, void (*destructor)(void*)) {
  if (key == 0)
    return EINVAL;
  if (!max_tls_entries) {
    // First time we're called, allocate entry table.
    max_tls_entries = 4;
    tls_entries = (struct entry_t*)malloc(max_tls_entries * sizeof(void *));
  }
  // Find empty spot.
  size_t entry = 0;
  for (; entry < num_tls_entries; entry++) {
    if (!tls_entries[entry].allocated) break;
  }
  if (entry == max_tls_entries) {
    // No empty spots, table full: double the table.
    max_tls_entries *= 2;
    tls_entries =
      (struct entry_t*)realloc(tls_entries, num_tls_entries * sizeof(void *));
  }
  if (entry == num_tls_entries) {
    // No empty spots, but table not full.
    num_tls_entries++;
  }
  struct entry_t* e = &tls_entries[entry];
  e->value = NULL;
  e->allocated = 1;
  // Key can't be zero.
  *key = (pthread_key_t)entry + 1;
  return 0;
}

int pthread_key_delete(pthread_key_t key) {
  if (key == 0 || key > num_tls_entries)
    return EINVAL;
  struct entry_t* e = &tls_entries[key - 1];
  if (!e->allocated)
    return EINVAL;
  e->value = NULL;
  e->allocated = 0;
  return 0;
}

void* pthread_getspecific(pthread_key_t key) {
  if (key == 0 || key > num_tls_entries)
    return NULL;
  struct entry_t* e = &tls_entries[key - 1];
  if (!e->allocated)
    return NULL;
  return (void*)e->value;
}

int pthread_setspecific(pthread_key_t key, const void* value) {
  if (key == 0 || key > num_tls_entries)
    return EINVAL;
  struct entry_t* e = &tls_entries[key - 1];
  if (!e->allocated)
    return EINVAL;
  e->value = value;
  return 0;
}

/*magic number to detect if we have not run yet*/
#define PTHREAD_ONCE_MAGIC_ID 0x13579BDF

int pthread_once(pthread_once_t* once_control, void (*init_routine)(void)) {
  if (*once_control != PTHREAD_ONCE_MAGIC_ID) {
    init_routine();
    *once_control = PTHREAD_ONCE_MAGIC_ID;
  }
  return 0;
}

int pthread_cond_wait(pthread_cond_t *cond, pthread_mutex_t *mutex) {
  return 0;
}

int pthread_cond_signal(pthread_cond_t *cond) {
  return 0;
}

int pthread_cond_broadcast(pthread_cond_t *cond) {
  return 0;
}

int pthread_cond_init(pthread_cond_t *__restrict x, const pthread_condattr_t *__restrict y) {
  return 0;
}

int pthread_cond_destroy(pthread_cond_t * x) {
  return 0;
}

int pthread_cond_timedwait(pthread_cond_t *__restrict x, pthread_mutex_t *__restrict y, const struct timespec *__restrict z) {
  return 0;
}

int pthread_atfork(void (*prepare)(void), void (*parent)(void), void (*child)(void)) {
  return 0;
}

// pthread_equal is defined as a macro in C, as a function for C++; undef it
// here so we define the function for C++ that links to us.
#undef pthread_equal

int pthread_equal(pthread_t t1, pthread_t t2) {
  return t1 == t2;
}
