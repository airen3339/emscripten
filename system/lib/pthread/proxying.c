/*
 * Copyright 2021 The Emscripten Authors.  All rights reserved.
 * Emscripten is available under two separate licenses, the MIT license and the
 * University of Illinois/NCSA Open Source License.  Both these licenses can be
 * found in the LICENSE file.
 */

#include <assert.h>
#include <emscripten/proxying.h>
#include <emscripten/threading.h>
#include <pthread.h>
#include <stdatomic.h>
#include <stdlib.h>
#include <string.h>

#include "em_task_queue.h"
#include "thread_mailbox.h"

struct em_proxying_queue {
  // Protects all accesses to em_task_queues, size, and capacity.
  pthread_mutex_t mutex;
  // `size` task queue pointers stored in an array of size `capacity`.
  em_task_queue** task_queues;
  int size;
  int capacity;
};

// The system proxying queue.
static em_proxying_queue system_proxying_queue = {
  .mutex = PTHREAD_MUTEX_INITIALIZER,
  .task_queues = NULL,
  .size = 0,
  .capacity = 0,
};

em_proxying_queue* emscripten_proxy_get_system_queue(void) {
  return &system_proxying_queue;
}

em_proxying_queue* em_proxying_queue_create(void) {
  // Allocate the new queue.
  em_proxying_queue* q = malloc(sizeof(em_proxying_queue));
  if (q == NULL) {
    return NULL;
  }
  *q = (em_proxying_queue){
    .mutex = PTHREAD_MUTEX_INITIALIZER,
    .task_queues = NULL,
    .size = 0,
    .capacity = 0,
  };
  return q;
}

void em_proxying_queue_destroy(em_proxying_queue* q) {
  assert(q != NULL);
  assert(q != &system_proxying_queue && "cannot destroy system proxying queue");

  pthread_mutex_destroy(&q->mutex);
  for (int i = 0; i < q->size; i++) {
    em_task_queue_destroy(q->task_queues[i]);
  }
  free(q->task_queues);
  free(q);
}

// Not thread safe. Returns NULL if there are no tasks for the thread.
static em_task_queue* get_tasks_for_thread(em_proxying_queue* q,
                                           pthread_t thread) {
  assert(q != NULL);
  for (int i = 0; i < q->size; i++) {
    if (pthread_equal(q->task_queues[i]->thread, thread)) {
      return q->task_queues[i];
    }
  }
  return NULL;
}

// Not thread safe.
static em_task_queue* get_or_add_tasks_for_thread(em_proxying_queue* q,
                                                  pthread_t thread) {
  em_task_queue* tasks = get_tasks_for_thread(q, thread);
  if (tasks != NULL) {
    return tasks;
  }
  // There were no tasks for the thread; initialize a new em_task_queue. If
  // there are not enough queues, allocate more.
  if (q->size == q->capacity) {
    int new_capacity = q->capacity == 0 ? 1 : q->capacity * 2;
    em_task_queue** new_task_queues =
      realloc(q->task_queues, sizeof(em_task_queue*) * new_capacity);
    if (new_task_queues == NULL) {
      return NULL;
    }
    q->task_queues = new_task_queues;
    q->capacity = new_capacity;
  }
  // Initialize the next available task queue.
  tasks = em_task_queue_create(thread);
  if (tasks == NULL) {
    return NULL;
  }
  q->task_queues[q->size++] = tasks;
  return tasks;
}

void emscripten_proxy_execute_queue(em_proxying_queue* q) {
  assert(q != NULL);
  assert(pthread_self());

  // Recursion guard to avoid infinite recursion when we arrive here from the
  // pthread_lock call below that executes the system queue. The per-task_queue
  // recursion lock can't catch these recursions because it can only be checked
  // after the lock has been acquired.
  static _Thread_local int executing_system_queue = 0;
  int is_system_queue = q == &system_proxying_queue;
  if (is_system_queue) {
    if (executing_system_queue) {
      return;
    }
    executing_system_queue = 1;
  }

  pthread_mutex_lock(&q->mutex);
  em_task_queue* tasks = get_tasks_for_thread(q, pthread_self());
  pthread_mutex_unlock(&q->mutex);

  if (tasks != NULL && !tasks->processing) {
    // Found the task queue and it is not already being processed; process it.
    em_task_queue_execute(tasks);
  }

  if (is_system_queue) {
    executing_system_queue = 0;
  }
}

int emscripten_proxy_async(em_proxying_queue* q,
                           pthread_t target_thread,
                           void (*func)(void*),
                           void* arg) {
  assert(q != NULL);
  pthread_mutex_lock(&q->mutex);
  em_task_queue* tasks = get_or_add_tasks_for_thread(q, target_thread);
  pthread_mutex_unlock(&q->mutex);
  if (tasks == NULL) {
    return 0;
  }

  return em_task_queue_send(tasks, (task){func, arg});
}

struct em_proxying_ctx {
  // The user-provided function and argument.
  void (*func)(em_proxying_ctx*, void*);
  void* arg;
  // Set `done` to 1 and signal the condition variable once the proxied task is
  // done.
  int done;
  pthread_mutex_t mutex;
  pthread_cond_t cond;
};

static void em_proxying_ctx_init(em_proxying_ctx* ctx,
                                 void (*func)(em_proxying_ctx*, void*),
                                 void* arg) {
  *ctx = (em_proxying_ctx){.func = func,
                           .arg = arg,
                           .done = 0,
                           .mutex = PTHREAD_MUTEX_INITIALIZER,
                           .cond = PTHREAD_COND_INITIALIZER};
}

static void em_proxying_ctx_deinit(em_proxying_ctx* ctx) {
  pthread_mutex_destroy(&ctx->mutex);
  pthread_cond_destroy(&ctx->cond);
}

void emscripten_proxy_finish(em_proxying_ctx* ctx) {
  pthread_mutex_lock(&ctx->mutex);
  ctx->done = 1;
  pthread_mutex_unlock(&ctx->mutex);
  pthread_cond_signal(&ctx->cond);
}

// Helper for wrapping the call with ctx as a `void (*)(void*)`.
static void call_with_ctx(void* p) {
  em_proxying_ctx* ctx = (em_proxying_ctx*)p;
  ctx->func(ctx, ctx->arg);
}

int emscripten_proxy_sync_with_ctx(em_proxying_queue* q,
                                   pthread_t target_thread,
                                   void (*func)(em_proxying_ctx*, void*),
                                   void* arg) {
  assert(!pthread_equal(target_thread, pthread_self()) &&
         "Cannot synchronously wait for work proxied to the current thread");
  em_proxying_ctx ctx;
  em_proxying_ctx_init(&ctx, func, arg);
  if (!emscripten_proxy_async(q, target_thread, call_with_ctx, &ctx)) {
    return 0;
  }
  pthread_mutex_lock(&ctx.mutex);
  while (!ctx.done) {
    pthread_cond_wait(&ctx.cond, &ctx.mutex);
  }
  pthread_mutex_unlock(&ctx.mutex);
  em_proxying_ctx_deinit(&ctx);
  return 1;
}

// Helper for signaling the end of the task after the user function returns.
static void call_then_finish(em_proxying_ctx* ctx, void* arg) {
  task* t = (task*)arg;
  t->func(t->arg);
  emscripten_proxy_finish(ctx);
}

int emscripten_proxy_sync(em_proxying_queue* q,
                          pthread_t target_thread,
                          void (*func)(void*),
                          void* arg) {
  task t = {func, arg};
  return emscripten_proxy_sync_with_ctx(q, target_thread, call_then_finish, &t);
}

// Helper struct for organizing a proxied call and its callback on the original
// thread.
struct callback {
  em_proxying_queue* q;
  pthread_t caller_thread;
  void* (*func)(void*);
  void* arg;
  void (*callback)(void* arg, void* result);
  void* callback_arg;
  void* result;
};

// Free the callback info on the same thread it was originally allocated on.
// This may be more efficient.
static void call_callback_then_free(void* arg) {
  struct callback* info = (struct callback*)arg;
  info->callback(info->callback_arg, info->result);
  free(arg);
}

static void call_then_schedule_callback(void* arg) {
  struct callback* info = (struct callback*)arg;
  info->result = info->func(info->arg);
  if (!emscripten_proxy_async(
        info->q, info->caller_thread, call_callback_then_free, arg)) {
    // No way to gracefully report that we failed to schedule the callback, so
    // abort.
    abort();
  }
}

int emscripten_proxy_async_with_callback(em_proxying_queue* q,
                                         pthread_t target_thread,
                                         void* (*func)(void*),
                                         void* arg,
                                         void (*callback)(void* arg,
                                                          void* result),
                                         void* callback_arg) {
  struct callback* info = malloc(sizeof(*info));
  if (info == NULL) {
    return 0;
  }
  *info = (struct callback){
    .q = q,
    .caller_thread = pthread_self(),
    .func = func,
    .arg = arg,
    .callback = callback,
    .callback_arg = callback_arg,
    .result = NULL,
  };
  return emscripten_proxy_async(
    q, target_thread, call_then_schedule_callback, info);
}
