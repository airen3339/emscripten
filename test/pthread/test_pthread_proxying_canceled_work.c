#include <assert.h>
#include <emscripten/proxying.h>
#include <pthread.h>
#include <stdio.h>

em_proxying_queue* queue;

void explode(void* arg) { assert(0 && "the work should not be run!"); }

void set_flag(void* flag) {
  // Schedule the flag to be set on the next turn of the event loop so that we
  // can be sure cleanup has finished first. We need to use EM_ASM and JS here
  // because this code needs to run after the thread runtime has exited.

  // clang-format off
  EM_ASM({setTimeout(() => Atomics.store(HEAP32, $0 >> 2, 1))}, flag);
  // clang-format on
}

// Used to call `set_flag` on thread exit or cancellation.
pthread_key_t dtor_key;

void* cancel_self(void* canceled) {
  pthread_setspecific(dtor_key, canceled);
  pthread_cancel(pthread_self());
  pthread_testcancel();
  assert(0 && "thread should have been canceled!");
  return NULL;
}

void* exit_self(void* exited) {
  pthread_setspecific(dtor_key, exited);
  pthread_exit(NULL);
  assert(0 && "thread should have exited!");
  return NULL;
}

void test_cancel_then_proxy() {
  printf("testing cancel followed by proxy\n");

  pthread_t thread;
  _Atomic int canceled = 0;
  pthread_create(&thread, NULL, cancel_self, &canceled);

  // Wait for the thread to be canceled.
  while (!canceled) {
  }

  // Proxying work to the thread should return an error.
  int ret = emscripten_proxy_sync(queue, thread, explode, NULL);
  assert(ret == 0);

  pthread_join(thread, NULL);
}

void test_exit_then_proxy() {
  printf("testing exit followed by proxy\n");

  pthread_t thread;
  _Atomic int exited = 0;
  pthread_create(&thread, NULL, exit_self, &exited);

  // Wait for the thread to exit.
  while (!exited) {
  }

  // Proxying work to the thread should return an error.
  int ret = emscripten_proxy_sync(queue, thread, explode, NULL);
  assert(ret == 0);

  pthread_join(thread, NULL);
}

int main() {
  queue = em_proxying_queue_create();
  pthread_key_create(&dtor_key, set_flag);

  test_cancel_then_proxy();
  test_exit_then_proxy();

  em_proxying_queue_destroy(queue);

  printf("done\n");
}
