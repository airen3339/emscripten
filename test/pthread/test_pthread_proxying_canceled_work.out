testing cancel followed by proxy
testing exit followed by proxy
done
